pkg/execution/state/redis_state/queue_test.go

```diff
@@ -1158,6 +1158,243 @@ func TestQueuePartitionReprioritize(t *testing.T) {
 	})
 }
 
+func TestRequeueByJobID(t *testing.T) {
+	ctx := context.Background()
+	r := miniredis.RunT(t)
+
+	rc, err := rueidis.NewClient(rueidis.ClientOption{
+		InitAddress:  []string{r.Addr()},
+		DisableCache: true,
+	})
+	require.NoError(t, err)
+	defer rc.Close()
+
+	q := queue{
+		kg: defaultQueueKey,
+		r:  rc,
+		pf: func(ctx context.Context, item QueueItem) uint {
+			return PriorityMin
+		},
+		partitionConcurrencyGen: func(ctx context.Context, p QueuePartition) (string, int) {
+			return p.Queue(), 100
+		},
+	}
+
+	wsA, wsB := uuid.New(), uuid.New()
+
+	t.Run("Failure cases", func(t *testing.T) {
+
+		t.Run("It fails with a non-existent partition and job ID", func(t *testing.T) {
+			err := q.RequeueByJobID(ctx, "foo", "bar", time.Now().Add(5*time.Second))
+			require.NotNil(t, err)
+		})
+
+		t.Run("It fails with a non-existent job ID for an existing partition", func(t *testing.T) {
+			r.FlushDB()
+
+			jid := "yeee"
+			item := QueueItem{
+				ID:          jid,
+				WorkflowID:  wsA,
+				WorkspaceID: wsA,
+			}
+
+			_, err := q.EnqueueItem(ctx, item, time.Now().Add(time.Second))
+			require.NoError(t, err)
+
+			err = q.RequeueByJobID(ctx, wsA.String(), "no bruv", time.Now().Add(5*time.Second))
+			require.NotNil(t, err)
+		})
+
+		t.Run("It fails with a non-existent partition but an existing job ID", func(t *testing.T) {
+			r.FlushDB()
+
+			jid := "another"
+			item := QueueItem{
+				ID:          jid,
+				WorkflowID:  wsA,
+				WorkspaceID: wsA,
+			}
+
+			_, err := q.EnqueueItem(ctx, item, time.Now().Add(time.Second))
+			require.NoError(t, err)
+
+			err = q.RequeueByJobID(ctx, wsB.String(), jid, time.Now().Add(5*time.Second))
+			require.NotNil(t, err)
+		})
+
+		t.Run("It fails if the job is leased", func(t *testing.T) {
+			r.FlushDB()
+
+			jid := "leased"
+			item := QueueItem{
+				ID:          jid,
+				WorkflowID:  wsA,
+				WorkspaceID: wsA,
+			}
+
+			item, err := q.EnqueueItem(ctx, item, time.Now().Add(time.Second))
+			require.NoError(t, err)
+
+			partitions, err := q.PartitionPeek(ctx, true, time.Now().Add(5*time.Second), 10)
+			require.NoError(t, err)
+			require.Equal(t, 1, len(partitions))
+
+			// Lease
+			lid, err := q.Lease(ctx, *partitions[0], item, time.Second*10)
+			require.NoError(t, err)
+			require.NotNil(t, lid)
+
+			err = q.RequeueByJobID(ctx, wsB.String(), jid, time.Now().Add(5*time.Second))
+			require.NotNil(t, err)
```

Comment:
non-blocking:
testing the actual message matches would be good in the future so we know exactly which error was hit, and have more confidence in the error handling.

the current code just mean any error, and is not necessarily accurate for testing specific failure scenarios.
same for the other error matches as well.